packages/utils/src/schema/retrieveSchema.ts

```diff
@@ -179,33 +214,48 @@ export function resolveReference<T = any, S extends StrictRJSFSchema = RJSFSchem
   schema: S,
   rootSchema: S,
   expandAllBranches: boolean,
+  recurseList: string[],
   formData?: T
 ): S[] {
-  // Drop the $ref property of the source schema.
-  const { $ref, ...localSchema } = schema;
-  // Retrieve the referenced schema definition.
-  const refSchema = findSchemaDefinition<S>($ref, rootSchema);
-  // Update referenced schema definition with local schema properties.
-  return retrieveSchemaInternal<T, S, F>(
-    validator,
-    { ...refSchema, ...localSchema },
-    rootSchema,
-    formData,
-    expandAllBranches
-  );
+  const updatedSchema = resolveAllReferences<S>(schema, rootSchema, recurseList);
+  if (updatedSchema !== schema) {
+    // Only call this if the schema was actually changed by the `resolveAllReferences()` function
+    return retrieveSchemaInternal<T, S, F>(
+      validator,
+      updatedSchema,
+      rootSchema,
+      formData,
+      expandAllBranches,
+      recurseList
+    );
+  }
+  return [schema];
 }
 
-/** Resolves all references within a schema's properties and array items.
+/** Resolves all references within the schema itself as well as any of its properties and array items.
  *
  * @param schema - The schema for which resolving all references is desired
  * @param rootSchema - The root schema that will be forwarded to all the APIs
- * @returns - given schema will all references resolved
+ * @param recurseList - List of $refs already resolved to prevent recursion
+ * @returns - given schema will all references resolved or the original schema if no internal `$refs` were resolved
  */
-export function resolveAllReferences<S extends StrictRJSFSchema = RJSFSchema>(schema: S, rootSchema: S): S {
+export function resolveAllReferences<S extends StrictRJSFSchema = RJSFSchema>(
+  schema: S,
+  rootSchema: S,
+  recurseList: string[]
+): S {
+  if (!isObject(schema)) {
+    return schema;
+  }
   let resolvedSchema: S = schema;
   // resolve top level ref
   if (REF_KEY in resolvedSchema) {
     const { $ref, ...localSchema } = resolvedSchema;
+    // Check for a recursive reference and stop the loop
+    if (recurseList.includes($ref!)) {
+      return resolvedSchema;
```

Comment:
This update here seems to have broken precompiled schemas. If the reference is not resolved for all schemas then `createAjvInstance` will error with `MissingRefError` for that schema that still has `$ref`.

Here is a simple example schema
```
{
  "anyOf": [
    {
      "title": "First method of identification",
      "properties": {
        "firstName": {
          "type": "string",
          "title": "First name"
        },
        "lastName": {
          "$ref": "#/definitions/test"
        }
      }
    },
    {
      "title": "Second method of identification",
      "properties": {
        "idCode": {
          "$ref": "#/definitions/test"
        }
      }
    }
  ],
  "definitions": {
    "test": {
      "type": "string"
    }
  }
}
```
and then the output from running `compileSchemaValidators`

```
if (schOrEnv === undefined) throw new MissingRefError(it.opts.uriResolver, baseId, $ref)
                                      ^
MissingRefError: can't resolve reference #/definitions/test from id -284587a1

  missingRef: '-284587a1#/definitions/test',
  missingSchema: '-284587a1'
```